### PR TITLE
feat: decimal128 type in common

### PR DESCRIPTION
**Pull request by QuenKar**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

implement `decimal128` type in `common`.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#2682

### CODE REVIEW

**Review by killme2008 — COMMENTED** [review]
Can we add some sqlness cases for decimal?

**Review by killme2008 — APPROVED** [review]
LGTM. After merging this PR, we can support decimal type in SQL and gRPC protocls etc.

File: src/common/decimal/src/decimal128.rs [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+    /// Create a new Decimal128 from i128, precision and scale.
+    pub fn new(value: i128, precision: u8, scale: i8) -> Self {
+        Self {
+            value,
+            precision,
+            scale,
+        }
+    }
+
+    pub fn try_new_decimal128(value: i128, precision: u8, scale: i8) -> error::Result<Self> {
```
Comment (zhongzc) [inline src/common/decimal/src/decimal128.rs:63]:
```suggestion
    pub fn try_new(value: i128, precision: u8, scale: i8) -> error::Result<Self> {
```

File: src/common/decimal/src/decimal128.rs [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+#[derive(Debug, Default, Eq, Copy, Clone, Serialize, Deserialize)]
+pub struct Decimal128 {
+    value: i128,
+    precision: u8,
+    scale: i8,
+}
+
+impl Decimal128 {
+    /// Create a new Decimal128 from i128, precision and scale.
+    pub fn new(value: i128, precision: u8, scale: i8) -> Self {
```
Comment (zhongzc) [inline src/common/decimal/src/decimal128.rs:55]:
```suggestion
    pub fn new_unchecked(value: i128, precision: u8, scale: i8) -> Self {
```

File: src/common/decimal/src/decimal128.rs [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+// Two decimal values can be compared if they have the same precision and scale.
+impl PartialOrd for Decimal128 {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        if self.precision == other.precision && self.scale == other.scale {
+            return self.value.partial_cmp(&other.value);
+        }
+        None
+    }
+}
```
Comment (zhongzc) [inline src/common/decimal/src/decimal128.rs:109]: Not sure if this behavior is expected. Rest LTGM

**Review by zhongzc — APPROVED** [review]

File: src/common/decimal/src/decimal128.rs (hunk 3 above)
Comment (QuenKar) [reply src/common/decimal/src/decimal128.rs:109]: this behavior is the same as arrow-datafusion [PartialEq](https://github.com/apache/arrow-datafusion/blob/b54990d7aaab38163d9666ffa4304f9e3060fd2d/datafusion/common/src/scalar.rs#L153-L155)  and [PartialOrd](https://github.com/apache/arrow-datafusion/blob/b54990d7aaab38163d9666ffa4304f9e3060fd2d/datafusion/common/src/scalar.rs#L257-L264)

**Pull request merged**